src/simulation/dynamics/_GeneralModuleFiles/extendedStateVector.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ ISC License
+
+ Copyright (c) 2021, Autonomous Vehicle Systems Lab, University of Colorado at Boulder
+
+ Permission to use, copy, modify, and/or distribute this software for any
+ purpose with or without fee is hereby granted, provided that the above
+ copyright notice and this permission notice appear in all copies.
+
+ THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+#include "extendedStateVector.h"
+
+ExtendedStateVector ExtendedStateVector::fromStates(const std::vector<DynamicObject *>& dynPtrs)
+{
+    return fromStateData(dynPtrs, [](const StateData& data){return data.getState();});
+}
+
+ExtendedStateVector ExtendedStateVector::fromStateDerivs(const std::vector<DynamicObject *>& dynPtrs)
+{
+    return fromStateData(dynPtrs, [](const StateData& data){return data.getStateDeriv();});
+}
+
+auto ExtendedStateVector::map(
+    std::function<Eigen::MatrixXd(const size_t &, const std::string &, const Eigen::MatrixXd &)> functor) const -> ExtendedStateVector
+{
+    ExtendedStateVector result;
+    result.reserve(this->size());
+    
+    for (auto&& pair : *this)
+    {
+        // pair.first.first -> index of the dynamic object to which the state belongs
+        // pair.first.second -> string that defines the state
+        // pair.second -> Eigen::MatrixXd with the state currently stored in ExtendedStateVector
+        result.emplace(pair.first, functor(pair.first.first, pair.first.second, pair.second));
+    }
+    
+    return result;
+}
+
+
+void ExtendedStateVector::apply( 
+        std::function<void(const size_t&, const std::string&, const Eigen::MatrixXd&)> functor 
+    ) const
+{
+    for (auto&& pair : *this)
+    {
+        // pair.first.first -> index of the dynamic object to which the state belongs
+        // pair.first.second -> string that defines the state
+        // pair.second -> Eigen::MatrixXd with the state currently stored in ExtendedStateVector
+
+        functor(pair.first.first, pair.first.second, pair.second);
+    }
+}
+
+
+void ExtendedStateVector::mapInPlace( 
+        std::function<Eigen::MatrixXd(const size_t&, const std::string&, const Eigen::MatrixXd&)> functor 
```

Comment:
Try passing `Eigen::MatrixXd&` and modify this value directly instead of passing a const and then copying. Will save a copy operation.